cdc/sink/codec/csv/csv_message.go

```diff
@@ -0,0 +1,218 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package csv
+
+import (
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/parser/mysql"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/pkg/config"
+	cerror "github.com/pingcap/tiflow/pkg/errors"
+)
+
+const (
+	insertOperation = "I"
+	deleteOperation = "D"
+	updateOperation = "U"
+)
+
+type csvMessage struct {
+	csvConfig  *config.CSVConfig
+	opType     string
+	tableName  string
+	schemaName string
+	commitTs   uint64
+	columns    []any
+	newRecord  bool
+}
+
+func newCSVMessage(config *config.CSVConfig) *csvMessage {
+	return &csvMessage{
+		csvConfig: config,
+		newRecord: true,
+	}
+}
+
+func (c *csvMessage) encode() []byte {
+	strBuilder := new(strings.Builder)
+	c.formatValue(c.opType, strBuilder)
+	c.formatValue(c.tableName, strBuilder)
+	c.formatValue(c.schemaName, strBuilder)
+	if c.csvConfig.IncludeCommitTs {
+		c.formatValue(c.commitTs, strBuilder)
+	}
+	for _, col := range c.columns {
+		c.formatValue(col, strBuilder)
+	}
+	strBuilder.WriteString(c.csvConfig.Terminator)
+	return []byte(strBuilder.String())
+}
+
+func (c *csvMessage) formatWithQuotes(value string, strBuilder *strings.Builder) {
+	quote := c.csvConfig.Quote
+
+	strBuilder.WriteString(quote)
+	strBuilder.WriteString(strings.ReplaceAll(value, quote, quote+quote))
+	strBuilder.WriteString(quote)
+}
+
+func (c *csvMessage) formatWithEscapes(value string, strBuilder *strings.Builder) {
+	lastPos := 0
+	delimiter := c.csvConfig.Delimiter
+
+	for i := 0; i < len(value); i++ {
+		ch := value[i]
+		isDelimiterStart := strings.HasPrefix(value[i:], delimiter)
+		if ch == config.CR || ch == config.LF || ch == config.Backslash || isDelimiterStart {
```

Comment:
Please add some comments to explain what you are doing.

Reply:
Done.